src/app/internal/guiapp.cpp

```diff
@@ -232,19 +232,29 @@ void GuiApp::perform()
             m->onDelayedInit();
         }
 
-        startupScenario()->runOnSplashScreen();
+        const auto finalizeStartup = [this, obj]() {
+            if (splashScreen) {
+                splashScreen->close();
+                delete splashScreen;
+            }
 
-        if (splashScreen) {
-            splashScreen->close();
-            delete splashScreen;
-        }
+            // The main window must be shown at this point so KDDockWidgets can read its size correctly
+            // and scale all sizes properly. https://github.com/musescore/MuseScore/issues/21148
+            QQuickWindow* w = dynamic_cast<QQuickWindow*>(obj);
+            w->setVisible(true);
 
-        // The main window must be shown at this point so KDDockWidgets can read its size correctly
-        // and scale all sizes properly. https://github.com/musescore/MuseScore/issues/21148
-        QQuickWindow* w = dynamic_cast<QQuickWindow*>(obj);
-        w->setVisible(true);
+            startupScenario()->runAfterSplashScreen();
+        };
 
-        startupScenario()->runAfterSplashScreen();
+        const muse::ProgressPtr progress = startupScenario()->splashScreenProgress();
```

Comment:
I think we could make runOnSplashScreen return a promise and then subscribe to it here.
Alternatively, we can keep progress as a separate method but explicitly call runOnSplashScreen.
Right now it’s not obvious that progress->start() starts the process.